src/typechecker.rs

```diff
@@ -2380,8 +2380,12 @@ pub fn typecheck_expression(
                 .find_struct_in_scope(0, "Array")
                 .expect("internal error: Array builtin definition not found");
 
-            let type_id =
-                project.find_or_add_type_id(Type::GenericInstance(array_struct_id, vec![inner_ty]));
+            // FIXME: Make sure type_hint is of type Array<T>
```

Comment:
My suggestion is to do something similar to `unify_with_type_hint` instead of the current "either the type hint or the known type", this allows things like generics to resolve correctly too.